cloudsql/sqlserver/database-sql/cloudsql.go

```diff
@@ -0,0 +1,306 @@
+func (app *App) indexHandler(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		if err := showTotals(w, r, app); err != nil {
+			log.Printf("showTotals: %v", err)
+			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+		}
+	case "POST":
+		if err := saveVote(w, r, app); err != nil {
+			log.Printf("saveVote: %v", err)
+			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+		}
+	default:
+		http.Error(w, fmt.Sprintf("HTTP Method %s Not Allowed", r.Method), http.StatusMethodNotAllowed)
+	}
+}
+
+func main() {
+
+	var err error
+
+	parsedTemplate, err = template.ParseFiles("templates/index.html")
+	if err != nil {
+		log.Fatalf("unable to parse template file: %s", err)
+	}
+
+	app := &App{}
+
+	// If the optional DB_TCP_HOST environment variable is set, it contains
+	// the IP address and port number of a TCP connection pool to be created,
+	// such as "127.0.0.1:1433". If DB_TCP_HOST is not set, a Unix socket
+	// connection pool will be created instead.
+	if os.Getenv("DB_TCP_HOST") != "" {
+		app.db, err = initTcpConnectionPool()
+		if err != nil {
+			log.Fatalf("initTcpConnectionPool: unable to connect: %s", err)
+		}
+	} else {
+		app.db, err = initSocketConnectionPool()
+		if err != nil {
+			log.Fatalf("initSocketConnectionPool: unable to connect: %s", err)
+		}
+	}
+
+	// Drop the votes table if it already exists.
+	if _, err = app.db.Exec(`DROP TABLE IF EXISTS votes;`); err != nil {
+		log.Fatalf("DB.Exec: unable to drop votes table: %s", err)
+	}
+	// Create the votes table.
+	if _, err = app.db.Exec(`CREATE TABLE votes
+	( vote_id int IDENTITY(1,1) PRIMARY KEY, time_cast DATETIME NOT NULL,
+	candidate CHAR(6) NOT NULL );`); err != nil {
+		log.Fatalf("DB.Exec: unable to create votes table: %s", err)
+	}
+
+	http.HandleFunc("/", app.indexHandler)
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
+	}
+
+	log.Printf("Listening on port %s", port)
+	if err := http.ListenAndServe(":"+port, nil); err != nil {
+		log.Fatal(err)
+	}
+
+}
+
+// recentVotes returns a slice of the last 5 votes cast.
+func recentVotes(app *App) ([]vote, error) {
+	var votes []vote
+	rows, err := app.db.Query(`SELECT TOP 5 candidate, time_cast FROM votes ORDER BY time_cast DESC`)
+	if err != nil {
+		return votes, fmt.Errorf("DB.Query: %v", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		nextVote := vote{}
+		err := rows.Scan(&nextVote.Candidate, &nextVote.VoteTime)
+		if err != nil {
+			return votes, fmt.Errorf("Rows.Scan: %v", err)
+		}
+		votes = append(votes, nextVote)
+	}
+	return votes, nil
+}
+
+// currentTotals returns a templateData structure for populating the web page.
+func currentTotals(app *App) (templateData, error) {
+	// get total votes for each candidate
+	var tabVotes, spaceVotes uint
+	err := app.db.QueryRow(`SELECT count(vote_id) FROM votes WHERE candidate='TABS'`).Scan(&tabVotes)
+	if err != nil {
+		return templateData{}, fmt.Errorf("DB.QueryRow: %v", err)
```

Comment:
Can return a pointer to avoid having to be able to switch this to `nil, fmt.Errorf(...)`. The value shouldn't be used when the error is not nil.

Reply:
Thanks, addressed in this commit: https://github.com/GoogleCloudPlatform/golang-samples/pull/1390/commits/849f3f7f099bc0f6e7d8e0d0e25f74beeb0bcb33